cmd/kola/etcdtests/util.go

```diff
@@ -0,0 +1,158 @@
+package etcdtests
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/coreos/mantle/platform"
+)
+
+// run etcd on each cluster machine
+func startEtcd(cluster platform.Cluster) error {
+	for i, m := range cluster.Machines() {
+		etcdStart := "sudo systemctl start etcd.service"
+		_, err := m.SSH(etcdStart)
+		if err != nil {
+			return fmt.Errorf("start etcd.service on %v failed: %s", m.IP(), err)
+		}
+		fmt.Fprintf(os.Stderr, "etcd instance%d started\n", i)
+	}
+	return nil
+}
+
+// stop etcd on each cluster machine
+func stopEtcd(cluster platform.Cluster) error {
+	for i, m := range cluster.Machines() {
+		// start etcd instance
+		etcdStop := "sudo systemctl stop etcd.service"
+		_, err := m.SSH(etcdStop)
+		if err != nil {
+			return fmt.Errorf("stop etcd.service on %v failed: %s", m.IP(), err)
+		}
+		fmt.Fprintf(os.Stderr, "etcd instance%d stopped\n", i)
+	}
+	return nil
+}
+
+type Key struct {
+	Node struct {
+		Value string `json:"value"`
+	} `json:"node"`
+}
+
+func setKey(cluster platform.Cluster, key, value string) error {
+	cmd := cluster.NewCommand("curl", "-L", "http://10.0.0.3:4001/v2/keys/"+key, "-XPUT", "-d", "value="+value)
+	b, err := cmd.Output()
+	if err != nil {
+		return fmt.Errorf("Error setting value to cluster: %v", err)
+	}
+	fmt.Fprintf(os.Stderr, "%s\n", b)
+
+	var k Key
+	err = json.Unmarshal(b, &k)
+	if err != nil {
+		return err
+	}
+	if k.Node.Value != value {
+		fmt.Errorf("etcd key not set correctly")
+	}
+	return nil
+}
+
+func getKey(cluster platform.Cluster, key string) (string, error) {
+	cmd := cluster.NewCommand("curl", "-L", "http://10.0.0.3:4001/v2/keys/"+key)
+	b, err := cmd.Output()
+	if err != nil {
+		return "", fmt.Errorf("error getting old value from cluster: %v", err)
+	}
+	fmt.Fprintf(os.Stderr, "%s\n", b)
+
+	var k Key
+	err = json.Unmarshal(b, &k)
+	if err != nil {
+		return "", err
+	}
+
+	return k.Node.Value, nil
+}
+
+// replace default binary for etcd.service with internal version v
+func replaceEtcdBin(cluster platform.Cluster, oldv, newv string) error {
+	// escape file path forward-slashes for sed command
+	oldv = strings.Replace(oldv, "/", `\/`, -1)
+	newv = strings.Replace(newv, "/", `\/`, -1)
+
+	for _, m := range cluster.Machines() {
+		// first err collector
+		var err error
+		firstErr := func(b []byte, e error) {
```

Comment:
I've used this pattern in order to perform all cleanup operations even if one of them failed instead of coming up with a scheme to return all the errors if more than one thing failed.

In this case there isn't any reason to continue execution if something fails so the standard pattern of wrapping each in an if is more appropriate.